test/unit/optionalChaining.spec.ts

```diff
@@ -1,9 +1,130 @@
-import { optionalChainingNotSupported } from "../../src/transformation/utils/diagnostics";
 import * as util from "../util";
 
-test("Diagnostic optional chaining is not supported yet", () => {
+test.each(["null", "undefined", '{ foo: "foo" }'])("optional chaining (%p)", value => {
     util.testFunction`
-        let func = (value: number) => value != 0 ? {value} : undefined;
-        return func(1)?.value;
-    `.expectToHaveDiagnostics([optionalChainingNotSupported.code]);
+        const obj: any = ${value};
+        return obj?.foo;
+    `.expectToMatchJsResult();
+});
+
+test("long optional chain", () => {
+    util.testFunction`
+        const a = { b: { c: { d: { e: { f: "hello!"}}}}};
+        return a.b?.c?.d.e.f;
+    `.expectToMatchJsResult();
+});
+
+test.each(["undefined", "{}", "{ foo: {} }", "{ foo: {bar: 'baz'}}"])("nested optional chaining (%p)", value => {
+    util.testFunction`
+        const obj: { foo?: { bar?: string } } | undefined = ${value};
+        return obj?.foo?.bar;
+    `.expectToMatchJsResult();
+});
+
+test.each(["undefined", "{}", "{ foo: {} }", "{ foo: {bar: 'baz'}}"])(
+    "nested optional chaining combined with coalescing (%p)",
+    value => {
+        util.testFunction`
+        const obj: { foo?: { bar?: string } } | undefined = ${value};
+        return obj?.foo?.bar ?? "not found";
+    `.expectToMatchJsResult();
+    }
+);
+
+test.each(["[1, 2, 3, 4]", "undefined"])("optional array access (%p)", value => {
+    util.testFunction`
+        const arr: number[] | undefined = ${value};
+        return arr?.[2];
+    `.expectToMatchJsResult();
+});
+
+test("optional array access nested", () => {
+    util.testFunction`
```

Comment:
It might be good to have additional test cases here with undefined array elements.